Isosurfacing_3/include/CGAL/Default_gradients.h

```diff
@@ -0,0 +1,168 @@
+class Finite_difference_gradient {
+public:
+    typedef GeomTraits Geom_traits;
+    typedef typename Geom_traits::FT FT;
+    typedef typename Geom_traits::Point_3 Point;
+    typedef typename Geom_traits::Vector_3 Vector;
+
+public:
+    Finite_difference_gradient(const PointFunction& func, const FT delta = 0.001) : func(func), delta(delta) {}
+
+    Vector operator()(const Point& point) const {  // TODO
+        // compute the gradient by sampling the function with finite differences
+
+        const Point p0 = point + Vector(delta, 0, 0);
+        const Point p1 = point - Vector(delta, 0, 0);
+        const Point p2 = point + Vector(0, delta, 0);
+        const Point p3 = point - Vector(0, delta, 0);
+        const Point p4 = point + Vector(0, 0, delta);
+        const Point p5 = point - Vector(0, 0, delta);
+
+        const FT gx = (func(p0) - func(p1)) / (2 * delta);
+        const FT gy = (func(p2) - func(p3)) / (2 * delta);
+        const FT gz = (func(p4) - func(p5)) / (2 * delta);
+
+        return Vector(gx, gy, gz);
+    }
+
+private:
+    const PointFunction func;
+    FT delta;
+};
+
+/**
+ * \ingroup PkgIsosurfacing3Ref
+ *
+ * \brief Template class for a gradient that is stored in a `Cartesian_grid_3`.
+ *
+ * \details The gradient at any point is calculated using trilinear interpolation.
+ *
+ * \tparam GeomTraits the traits for this gradient.
+ */
+template <class GeomTraits>
+class Explicit_cartesian_grid_gradient {
+public:
+    typedef GeomTraits Geom_traits;
+    typedef typename Geom_traits::FT FT;
+    typedef typename Geom_traits::Point_3 Point;
+    typedef typename Geom_traits::Vector_3 Vector;
+
+    typedef std::shared_ptr<Cartesian_grid_3<Geom_traits>> Grid;
+
+public:
+    Explicit_cartesian_grid_gradient(const Grid& grid) : grid(grid) {}
+
+    Vector operator()(const Point& point) const {
+        // trilinear interpolation of stored gradients
+
+        const Bbox_3& bbox = grid->get_bbox();
+        const Vector& spacing = grid->get_spacing();
+
+        // calculate min index including border case
+        std::size_t min_i = (point.x() - bbox.xmin()) / spacing.x();
+        std::size_t min_j = (point.y() - bbox.ymin()) / spacing.y();
+        std::size_t min_k = (point.z() - bbox.zmin()) / spacing.z();
+        if (min_i == grid->xdim() - 1) {
+            min_i--;
+        }
+        if (min_j == grid->ydim() - 1) {
+            min_j--;
+        }
+        if (min_k == grid->zdim() - 1) {
+            min_k--;
+        }
+
+        const FT min_x = min_i * spacing.x() + bbox.xmin();
+        const FT min_y = min_j * spacing.y() + bbox.ymin();
+        const FT min_z = min_k * spacing.z() + bbox.zmin();
+
+        const FT f_i = (point.x() - min_x) / spacing.x();
+        const FT f_j = (point.y() - min_y) / spacing.y();
+        const FT f_k = (point.z() - min_z) / spacing.z();
+
+        const Vector g000 = grid->gradient(min_i + 0, min_j + 0, min_k + 0);
```

Comment:
In case we expect `FT` to be something else than `double` or `float`  we maybe should not add `0`  and we should write `const FT one(1);`

Reply:
min_(i | j | k) are of type std::size_t so this in not an issue here. But I can check that this doesn't occur elsewhere in the code.